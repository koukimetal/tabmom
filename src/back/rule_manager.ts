import {
    getAllTabs,
    MessageType,
    getRule,
    TabInfo,
    getCurrentTime,
    setCurrentTime,
    TimerMessage,
    UpdateRuleMessage,
    setRule,
} from '../proxy';
import { CronRule, TimeRangeType, SkipInfo, ClockConfig } from '../components/system/actions';

export class RuleManager {
    private readonly nowDay: number;
    private readonly nowMinutes: number;
    private readonly previousMinutes: number;
    private readonly tabs: TabInfo[];

    private constructor(nowDay: number, nowMinutes: number, previousMinutes: number, tabs: TabInfo[]) {
        this.nowDay = nowDay;
        this.nowMinutes = nowMinutes;
        this.previousMinutes = previousMinutes;
        this.tabs = tabs;
    }

    public static async getInstance(nowDay: number, nowMinutes: number, previousMinutes: number) {
        const allTabs = await getAllTabs();
        return new RuleManager(nowDay, nowMinutes, previousMinutes, allTabs);
    }

    public isSatisfyOptions(skipInfo?: SkipInfo, weekSetting?: boolean[]) {
        if (weekSetting) {
            if (!weekSetting[this.nowDay]) {
                return false;
            }
        }

        if (skipInfo) {
            const hitTab = this.tabs
                .filter(tab => !(skipInfo.ignorePinned && tab.pinned))
                .find(tab => tab.url.includes(skipInfo.match));
            if (hitTab) {
                return false;
            }
        }
        return true;
    }

    public async isShouldOpen(id: string, clockConfig: ClockConfig, skipInfo?: SkipInfo, weekSetting?: boolean[]) {
        if (clockConfig.type === TimeRangeType.ONCE) {
            const minutes = clockConfig.startTime;
            const isFitTime = this.previousMinutes < minutes && minutes <= this.nowMinutes;
            return isFitTime && this.isSatisfyOptions(skipInfo, weekSetting);
        } else if (clockConfig.type === TimeRangeType.ALL || clockConfig.type === TimeRangeType.MANY) {
            const current = (await getCurrentTime(id)) || clockConfig.period;
            if (clockConfig.type === TimeRangeType.ALL) {
                return this.isSatisfyOptions(skipInfo, weekSetting) && current <= 1;
            } else {
                // MANY
                const inRange = clockConfig.startTime <= this.nowMinutes && this.nowMinutes <= clockConfig.endTime;
                return this.isSatisfyOptions(skipInfo, weekSetting) && inRange && current <= 1;
            }
        }
    }

    public async updateNextCurrent(id: string, clockConfig: ClockConfig) {
        if (
            clockConfig.type !== TimeRangeType.ONCE &&
            (clockConfig.type === TimeRangeType.ALL || clockConfig.type === TimeRangeType.MANY)
        ) {
            const current = (await getCurrentTime(id)) || clockConfig.period;
            const nextCurrent = current <= 1 ? clockConfig.period : current - 1;
            await setCurrentTime(id, nextCurrent);
            const message: TimerMessage = { type: MessageType.TIMER, id, time: nextCurrent };
            chrome.runtime.sendMessage(message);
        }
    }

    public async handleOpenTab(rule: CronRule) {
        if (rule.oneTime) {
            const nextRule = Object.assign({}, rule, { active: false });
            const message: UpdateRuleMessage = {
                type: MessageType.UPDATE_RULE,
                rule: nextRule,
            };
            await setRule(nextRule);
            chrome.runtime.sendMessage(message);
        }
        chrome.tabs.create({ url: rule.url });
    }

    public handleRule = async (id: string) => {
        const rule = await getRule(id);
        const { clockConfig, skipInfo, weekSetting, active } = rule;
        if (!active) {
            return;
        }

        const shouldOpen = await this.isShouldOpen(id, clockConfig, skipInfo, weekSetting);
        await this.updateNextCurrent(id, clockConfig);

        if (shouldOpen) {
            await this.handleOpenTab(rule);
        }
    };
}
