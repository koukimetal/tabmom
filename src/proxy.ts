import { CronRule, SystemDate } from './components/system/actions';

const RULE_ORDER_KEY = 'rule_order';
const RULE_PREFIX = 'RULE_';
const CURRNET_PREFIX = 'CURRENT_';

export interface TabInfo {
    url: string;
    pinned: boolean;
}

export const setRule = (rule: CronRule) =>
    new Promise(resolve => {
        const key = RULE_PREFIX + rule.id;
        chrome.storage.local.set({ [key]: rule }, () => {
            resolve();
        });
    });

export const getRule = (id: string) =>
    new Promise<CronRule>(resolve => {
        const key = RULE_PREFIX + id;
        chrome.storage.local.get([key], values => {
            resolve(values[key]);
        });
    });

export const deleteRule = (id: string) =>
    new Promise<void>(resolve => {
        const key = RULE_PREFIX + id;
        chrome.storage.local.remove([key], () => {
            resolve();
        });
    });

export const setRuleOrder = (ruleOrder: string[]) =>
    new Promise(resolve => {
        chrome.storage.local.set({ [RULE_ORDER_KEY]: ruleOrder }, () => {
            resolve();
        });
    });

export const getRuleOrder = () =>
    new Promise<string[]>(resolve => {
        chrome.storage.local.get([RULE_ORDER_KEY], values => {
            resolve(values[RULE_ORDER_KEY]);
        });
    });

export const setCurrentTime = (id: string, time: number) =>
    new Promise<void>(resolve => {
        const key = CURRNET_PREFIX + id;
        chrome.storage.local.set({ [key]: time }, () => {
            resolve();
        });
    });

export const getCurrentTime = (id: string) =>
    new Promise<number>(resolve => {
        const key = CURRNET_PREFIX + id;
        chrome.storage.local.get([key], values => {
            resolve(values[key]);
        });
    });

export const deleteCurrentTime = (id: string) =>
    new Promise<void>(resolve => {
        const key = CURRNET_PREFIX + id;
        chrome.storage.local.remove(key, () => {
            resolve();
        });
    });

export const getAllTabs = () =>
    new Promise<TabInfo[]>(resolve => {
        chrome.tabs.query({}, tabs => {
            const tabArray: TabInfo[] = tabs.map(({ url, pinned }) => ({
                url,
                pinned,
            }));
            resolve(tabArray);
        });
    });

export enum MessageType {
    TIMER,
    UPDATE_RULE,
    UPDATE_NOW_DATE,
}

export interface TimerMessage {
    type: MessageType.TIMER;
    id: string;
    time: number;
}

export interface UpdateRuleMessage {
    type: MessageType.UPDATE_RULE;
    rule: CronRule;
}

export interface UpdateNowDateMessage {
    type: MessageType.UPDATE_NOW_DATE;
    date: SystemDate;
}

export type TabMomMessage = TimerMessage | UpdateRuleMessage | UpdateNowDateMessage;
