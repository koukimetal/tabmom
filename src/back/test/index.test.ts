jest.mock('../../shared', () => ({
    calculateNowMinutes: jest.fn().mockImplementation(() => 10),
}));
import { updateNowDate, alarmListner, AlarmName } from '../';
import { RuleManager } from '../rule_manager';
describe('test for back', () => {
    beforeEach(() => {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        (global as any).chrome = {
            runtime: {
                sendMessage: jest.fn(),
            },
        };
    });
    test('updateNowDate', () => {
        jest.unmock('../../proxy');
        updateNowDate(1, 2);
        expect((chrome.runtime.sendMessage as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
    });
    test('test index', async () => {
        const getRuleOrder = jest.fn().mockResolvedValue([]);
        require('../../proxy').getRuleOrder = getRuleOrder;
        jest.mock('../rule_manager');
        RuleManager.getInstance = jest.fn().mockResolvedValue({
            handleRule: jest.fn(),
        });
        const updatePreviousMinutes = jest.fn();
        await alarmListner({ name: AlarmName.MINUTE } as chrome.alarms.Alarm, 1, updatePreviousMinutes);
        expect(getRuleOrder.mock.calls.length).toEqual(1);
        expect(updatePreviousMinutes.mock.calls[0][0]).toEqual(10);
    });
});
