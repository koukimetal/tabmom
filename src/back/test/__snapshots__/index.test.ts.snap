// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test for back updateNowDate 1`] = `
Object {
  "date": Object {
    "nowDay": 2,
    "nowMinutes": 1,
  },
  "type": 2,
}
`;
